File: controllo_accessi_portale/static/src/js/badge_release.js
/** @odoo-module */
import publicWidget from "@web/legacy/js/public/public_widget";
import { onMounted, onWillStart, useState, Component, mount } from '@odoo/owl';
import { SelectMenu } from "@web/core/select_menu/select_menu";
import { useService } from "@web/core/utils/hooks";
import { _t } from "@web/core/l10n/translation";
import { getTemplate } from "@web/core/templates";

const { DateTime } = luxon;

class BadgeRelease extends Component {
    setup() {
        this.csrfToken = odoo.csrf_token;
        this.values = this.props.values || {};
        this.errors = this.props.errors || {};
        this.errorMessage = this.props.error_message || "";
        this.dataService = useService('dataService');

        this.texts = {
            issueBadge: _t("Issue Badge"),
            partnerList: _t("Partner List"),
            findInSystem: _t("Find in System"),
            existingPartner: _t("Existing Partner"),
            selectPartnerHelp: _t("Select a partner to pre-fill data or leave blank to create a new one."),
            personalDetails: _t("Personal Details"),
            lastName: _t("Last Name *"),
            firstName: _t("First Name *"),
            fiscalCode: _t("Fiscal Code *"),
            vatNumber: _t("VAT Number"),
            companyName: _t("Company Name / Entity *"),
            entityType: _t("Entity Type"),
            email: _t("Email"),
            mobile: _t("Mobile"),
            idNumber: _t("ID Number"),
            internalContact: _t("Internal Contact"),
            searchDomain: _t("Search Domain"),
            presentToday: _t("Present Today"),
            all: _t("All"),
            referencePerson: _t("Reference Person"),
            accessDetails: _t("Access Details"),
            workType: _t("Work Type *"),
            titleQualification: _t("Title / Qualification *"),
            assignBadge: _t("Assign Badge *"),
            startDate: _t("Start Date"),
            endDate: _t("End Date"),
            selectPartner: _t("Select a Partner..."),
            selectRepresentative: _t("Select a representative..."),
            selectBadge: _t("Select a badge..."),
            selectWorkInfo: _t("Select a work info type..."),
            selectTitle: _t("Select a title..."),
            fieldRequired: _t("This field is required.")
        };

        this.ca_persona = [];
        this.ca_persona_parent = [];
        this.tipo_ente_azienda = [];
        this.tipo_ente_azienda_hidden = [];
        this.work_info_type = [];
        this.titolo_persona = [];
        this.ca_ente_azienda = [];
        this.tags = [];

        this.store = {
            workInfosMap: {},
            entiMap: {},
            entiVatMap: {},
            titoliMap: {},
            personasMap: {},
            tagDomains: {},
        };

        this.state = useState({
            csrfToken: odoo.csrf_token,
            selectedPersonaId: null,
            caPersonaParentFiltered: [],
            availableTags: [],
            enteInterno: false,
            isPersonaInternal: false,
            submitted: false,

            formValues: {
                name: "", lastname: "", fiscalcode: "", freshman: "", email: "", mobile: "",
                ente_azienda: null, ca_ente_name: "", tipo_ente_azienda_id: null, vat: "",
                date_start: this.values.date_start ? DateTime.fromISO(this.values.date_start) : DateTime.local(),
                date_end: this.values.date_end ? DateTime.fromISO(this.values.date_end) : DateTime.local(),
                ca_tag_id: null, ref_domain: "present", work_info_type: null, ca_title: null,
                parent_id: null,
                ...this.values
            },
            errors: this.errors,
            errorMessage: this.errorMessage
        });

        onWillStart(async () => {
            const [initData, personas] = await Promise.all([
                this.dataService.loadBadgeReleaseInitData(),
                this.dataService.loadPersona()
            ]);

            const safeInit = initData || {};
            const safePersonas = personas || [];

            this.ca_persona = safePersonas;
            this.ca_persona_parent = safeInit.persona_parent || [];
            this.tipo_ente_azienda = safeInit.tipo_enti || [];
            this.tipo_ente_azienda_hidden = safeInit.tipo_enti_hidden_ids || [];
            this.work_info_type = safeInit.work_info_types || [];
            this.titolo_persona = safeInit.titoli || [];
            this.ca_ente_azienda = safeInit.enti_aziende || [];
            this.tags = safeInit.tags || [];
            this.store.tagDomains = safeInit.tag_domains || {};

            this.store.personasMap = Object.fromEntries(safePersonas.map(p => [p.id, p]));
            this.store.titoliMap = Object.fromEntries(this.titolo_persona.map(t => [t.id, t]));
            this.store.entiMap = Object.fromEntries(this.ca_ente_azienda.map(e => [e.id, e]));

            this.ca_ente_azienda.forEach(e => { if (e.vat) this.store.entiVatMap[e.vat] = e; });
            (safeInit.work_infos || []).forEach(w => {
                if (w.ca_persona_id) this.store.workInfosMap[w.ca_persona_id[0]] = w;
            });

            if (this.values.persona_id) this.state.selectedPersonaId = parseInt(this.values.persona_id);
            if (this.values.ca_work_info_type_id) this.state.formValues.work_info_type = parseInt(this.values.ca_work_info_type_id);
            if (this.values.ca_title_id) this.state.formValues.ca_title = parseInt(this.values.ca_title_id);
            if (this.values.ca_tag_id) this.state.formValues.ca_tag_id = parseInt(this.values.ca_tag_id);

            this.evalParentPresent();
            this.filterAvailableTagsSimple();
        });

        onMounted(() => {
            const staticLoader = document.getElementById('static_loader');
            if (staticLoader) staticLoader.remove();
        });
    }

    // --- Choice getters for SelectMenu ---

    get personaChoices() {
        return this.ca_persona.map(p => ({ value: p.id, label: p.display_name }));
    }

    get parentChoices() {
        return this.state.caPersonaParentFiltered.map(p => ({ value: p.id, label: p.display_name }));
    }

    get tagChoices() {
        return this.state.availableTags.map(t => ({ value: t.id, label: t.display_name }));
    }

    get workInfoChoices() {
        return this.work_info_type.map(w => ({ value: w.id, label: w.display_name }));
    }

    get titleChoices() {
        return this.titolo_persona.map(t => ({ value: t.id, label: t.display_name }));
    }

    // --- SelectMenu handlers (receive value directly, not event) ---

    onPersonaSelect(value) {
        const id = value || null;
        this.state.selectedPersonaId = id;
        this.state.formValues.persona_id = id;

        if (!id) {
            this.state.isPersonaInternal = false;
            this.filterAvailableTagsSimple();
            return;
        }
        const persona = this.store.personasMap[id];
        if (persona) this.populatePersona(persona);
    }

    onParentSelect(value) {
        this.state.formValues.parent_id = value || null;
    }

    onTagSelect(value) {
        const id = value || null;
        this.state.formValues.ca_tag_id = id;
        if (id) {
            const tag = this.tags.find(t => t.id === id);
            if (tag?.temp) {
                const now = DateTime.now().setZone('Europe/Rome');
                this.state.formValues.date_start = now;
                this.state.formValues.date_end = now.set({ hour: 19, minute: 30, second: 0 });
            }
        }
    }

    onWorkInfoSelect(value) {
        this.state.formValues.work_info_type = value || null;
    }

    onTitleSelect(value) {
        this.state.formValues.ca_title = value || null;
    }

    // --- Business logic ---

    populatePersona(persona) {
        const wInfo = this.store.workInfosMap[persona.id];
        let ente = null;
        if (persona.ca_ente_azienda_ids?.[0]) {
            ente = this.store.entiMap[persona.ca_ente_azienda_ids[0]];
        }

        this.state.isPersonaInternal = persona.is_internal || false;
        this.checkEnteInterno(ente);

        let dStart = this.state.formValues.date_start;
        let dEnd = this.state.formValues.date_end;
        if (wInfo) {
            if (wInfo.date_start) dStart = DateTime.fromISO(wInfo.date_start).set({ hour: 8, minute: 0, second: 0 });
            if (wInfo.date_end) dEnd = DateTime.fromISO(wInfo.date_end).set({ hour: 18, minute: 0, second: 0 });
        }

        Object.assign(this.state.formValues, {
            name: persona.name || "",
            lastname: persona.lastname || "",
            fiscalcode: persona.fiscalcode || "",
            freshman: persona.freshman || "",
            email: persona.email || "",
            mobile: persona.mobile || "",
            ente_azienda: ente?.id || "",
            ca_ente_name: ente?.name || "",
            tipo_ente_azienda_id: ente?.tipo_ente_azienda_id?.[0] || "",
            vat: ente?.vat || "",
            work_info_type: wInfo?.ca_work_info_type_id?.[0] || null,
            ca_title: wInfo?.ca_title_id?.[0] || null,
            date_start: dStart,
            date_end: dEnd
        });

        // SelectMenu reacts to state changes — no jQuery needed

        this.filterAvailableTagsSimple();
    }

    onVatChange(ev) {
        const vat = ev.target.value;
        this.state.formValues.vat = vat;
        const ente = this.store.entiVatMap[vat];
        if (ente) {
            Object.assign(this.state.formValues, {
                ca_ente_name: ente.name,
                ente_azienda: ente.id,
                tipo_ente_azienda_id: ente.tipo_ente_azienda_id?.[0]
            });
            this.checkEnteInterno(ente);
        }
    }

    checkEnteInterno(ente) {
        this.state.enteInterno = false;
        if (ente?.tipo_ente_azienda_id?.[0]) {
            if (this.tipo_ente_azienda_hidden.includes(ente.tipo_ente_azienda_id[0])) {
                this.state.enteInterno = true;
            }
        }
    }

    filterAvailableTagsSimple() {
        const persona = this.store.personasMap[this.state.selectedPersonaId];
        const domains = this.store.tagDomains;

        if (!persona || !persona.current_tag || persona.current_tag.length === 0) {
            this.state.availableTags = this.tags.filter(t => !t.in_use && !t.revoked);
            return;
        }

        const neededIds = persona.current_tag_temp === false
            ? (domains.temp || [])
            : (domains.visitor || []);

        this.state.availableTags = this.tags.filter(tag =>
            !tag.in_use &&
            !tag.revoked &&
            tag.ca_proprieta_tag_ids.some(id => neededIds.includes(id))
        );
    }

    filterAvailableTags() {
        const titleId = this.state.formValues.ca_title;
        const title = this.store.titoliMap[titleId];
        const persona = this.store.personasMap[this.state.selectedPersonaId];
        const domains = this.store.tagDomains;
        let neededIds = [];

        if (persona?.current_tag?.length) neededIds = domains.visitor;
        else if (title?.structured === false) neededIds = domains.temp;
        else if (persona?.is_internal) neededIds = domains.internal;
        else {
            this.state.availableTags = this.tags.filter(t => !t.in_use && !t.revoked);
            return;
        }

        if (neededIds && neededIds.length > 0) {
            this.state.availableTags = this.tags.filter(t =>
                !t.in_use && !t.revoked && t.ca_proprieta_tag_ids.some(id => neededIds.includes(id))
            );
        } else {
            this.state.availableTags = [];
        }
    }

    onRefDomainChange(ev) {
        this.state.formValues.ref_domain = ev.target.value;
        this.evalParentPresent();
    }

    evalParentPresent() {
        if (this.state.formValues.ref_domain === "present") {
            this.state.caPersonaParentFiltered = this.ca_persona_parent.filter(p => p.present === "yes");
        } else {
            this.state.caPersonaParentFiltered = this.ca_persona_parent || [];
        }
    }

    formatDateForInput(luxonDate) {
        if (!luxonDate || !luxonDate.isValid) return "";
        return luxonDate.toFormat("yyyy-MM-dd'T'HH:mm");
    }

    onDateStartChange(ev) {
        const val = ev.target.value;
        this.state.formValues.date_start = val ? DateTime.fromISO(val) : null;
    }

    onDateEndChange(ev) {
        const val = ev.target.value;
        this.state.formValues.date_end = val ? DateTime.fromISO(val) : null;
    }

    onNameChange(e) { this.state.formValues.name = e.target.value; }
    onLastnameChange(e) { this.state.formValues.lastname = e.target.value; }
    onEnteNameChange(e) { this.state.formValues.ca_ente_name = e.target.value; }
    OnTipoEnteChange(e) { this.state.formValues.tipo_ente_azienda_id = e.target.value; }
    OnFreshmanChange(e) { this.state.formValues.freshman = e.target.value; }

    onEmailChange(e) {
        this.state.formValues.email = e.target.value;
        const found = Object.values(this.store.personasMap).find(p => p.email === e.target.value);
        if (found) this.onPersonaSelect(found.id);
    }

    onMobileChange(e) { this.state.formValues.mobile = e.target.value; }

    onFiscalcodeChange(ev) {
        const val = ev.target.value;
        this.state.formValues.fiscalcode = val;
        const found = Object.values(this.store.personasMap).find(p => p.fiscalcode === val);
        if (found) this.onPersonaSelect(found.id);
    }

    onSubmitClick(e) {
        this.state.submitted = true;
        const form = document.querySelector("form");
        if (form) {
            form.classList.add('was-validated');

            const requiredMissing =
                !this.state.formValues.work_info_type ||
                !this.state.formValues.ca_title ||
                !this.state.formValues.ca_tag_id ||
                (!this.state.isPersonaInternal && !this.state.formValues.parent_id);

            if (!form.checkValidity() || requiredMissing) {
                e.preventDefault();
                e.stopPropagation();
                const invalid = form.querySelector(":invalid");
                if (invalid) invalid.scrollIntoView({ behavior: "smooth", block: "center" });
            }
        }
    }
}

BadgeRelease.components = { SelectMenu };
BadgeRelease.template = 'controllo_accessi_portale.BadgeRelease';
BadgeRelease.props = {
    values: { type: Object, optional: true },
    errors: { type: Object, optional: true },
    error_message: { type: String, optional: true },
};

publicWidget.registry.BadgeReleaseWidget = publicWidget.Widget.extend({
    selector: '#badge_release_app',

    start: function () {
        const serverData = window.odoo_badge_release_data || {};

        return mount(BadgeRelease, this.el, {
            getTemplate: getTemplate,
            props: {
                values: serverData.values || {},
                errors: serverData.errors || {},
                error_message: serverData.error_message || ""
            },
            env: Component.env,
            dev: odoo.debug,
        });
    }
});

export default BadgeRelease;

File: controllo_accessi_portale/controllers/badge_release.py
import datetime
import pytz
from pytz import UTC
from odoo import http, _
from odoo.http import request
from werkzeug.exceptions import Forbidden, NotFound
from .common import check_access_permission


class PortalBadgeRelease(http.Controller):

    def _get_user_timezone(self, env):
        tz_name = env.user.tz or env.context.get('tz') or 'UTC'
        try:
            return pytz.timezone(tz_name)
        except pytz.UnknownTimeZoneError:
            return UTC

    def _parse_portal_datetime(self, env, post, field_name):
        value = post.get(f'{field_name}_iso') or post.get(field_name)
        if not value:
            return None

        parsed = datetime.datetime.fromisoformat(value)
        if parsed.tzinfo is None:
            parsed = self._get_user_timezone(env).localize(parsed)

        return parsed.astimezone(UTC).replace(tzinfo=None)

    def _get_badge_release_personas(self, env, domain=None):
        fields = [
            'id', 'display_name', 'name', 'lastname', 'fiscalcode', 'email',
            'mobile', 'is_internal', 'freshman', 'ca_ente_azienda_ids',
            'current_tag', 'present',
        ]
        personas = env['ca.persona'].search_read(domain or [], fields=fields)
        current_tag_ids = [
            persona['current_tag'][0]
            for persona in personas
            if persona.get('current_tag')
        ]
        current_tag_map = {}
        if current_tag_ids:
            current_tag_map = {
                tag['id']: tag['temp']
                for tag in env['ca.tag_persona'].search_read(
                    [('id', 'in', current_tag_ids)],
                    ['temp'],
                )
            }
        for persona in personas:
            current_tag = persona.get('current_tag')
            persona['current_tag_temp'] = (
                current_tag_map.get(current_tag[0]) if current_tag else None
            )
        return personas

    def _get_existing_persona(self, env, vals):
        if not vals.get('persona_id'):
            return env['ca.persona']
        return env['ca.persona'].browse(vals['persona_id']).exists()

    def _populate_company_vals_from_persona(self, vals, persona):
        if not persona:
            return vals
        ente = persona.ca_ente_azienda_ids[:1]
        if not ente:
            return vals
        if not vals.get('ente_azienda'):
            vals['ente_azienda'] = ente.id
        if not vals.get('ca_ente_name'):
            vals['ca_ente_name'] = ente.name
        if not vals.get('tipo_ente_azienda_id'):
            vals['tipo_ente_azienda_id'] = ente.tipo_ente_azienda_id.id
        if not vals.get('vat'):
            vals['vat'] = ente.vat
        return vals

    def _get_badge_release_required_fields(self, persona):
        req_fields = [
            "lastname", "name", "fiscalcode", "ca_ente_name",
            "ca_work_info_type_id", "ca_title_id", "date_start",
            "date_end", "ca_tag_id",
        ]
        if not persona:
            req_fields.insert(3, "tipo_ente_azienda_id")
        if not (persona and persona.is_internal):
            req_fields.extend(["ref_domain", "parent_id"])
        return req_fields
    
    @http.route('/badge_release', auth='user', type='http', website=True)
    def badge_release_form(self, **post):
        user = request.env.user
        if not check_access_permission(user):
            raise NotFound()

        if post and request.httprequest.method == 'POST':
            return self._handle_badge_release_post(post)

        vals = {k: v for k, v in post.items()} if post else {}

        return request.render('controllo_accessi_portale.badge_release_view', {
            "errors": {}, "error_message": "", "values": vals
        })

    @http.route('/badge_release/submit', auth='user', type='http', website=True, methods=['POST'], csrf=False)
    def badge_release_submit(self, **kwargs):
        return self._handle_badge_release_post(kwargs)

    def _handle_badge_release_post(self, post):
        vals = {
            'vat': post['vat'], 'ca_ente_name': post['ca_ente_name'], 'fiscalcode': post['fiscalcode'],
            'lastname': post['lastname'], 'name': post['name'], 'freshman': post['freshman'],
            'email': post['email'], 'mobile': post['mobile'], 'ref_domain': post.get('ref_domain'),
        }

        date_start = self._parse_portal_datetime(request.env, post, 'date_start')
        date_end = self._parse_portal_datetime(request.env, post, 'date_end')
        if date_start:
            vals['date_start'] = date_start
        if date_end:
            vals['date_end'] = date_end

        for key in ['persona_id', 'tipo_ente_azienda_id', 'azienda', 'ca_work_info_type_id', 'ca_title_id', 'ca_tag_id', 'parent_id']:
            if post.get(key) and post[key] != "":
                clean_key = 'ente_azienda' if key == 'azienda' else key
                vals[clean_key] = int(post[key])

        persona = self._get_existing_persona(request.env, vals)
        vals = self._populate_company_vals_from_persona(vals, persona)
        req_fields = self._get_badge_release_required_fields(persona)

        errors = {
            field_name: 'missing'
            for field_name in req_fields
            if not vals.get(field_name)
        }
        if errors:
            return request.render('controllo_accessi_portale.badge_release_view', {
                "errors": errors, "error_message": _('Some required fields are empty.'), "values": post
            })

        add_doc = not vals.get("persona_id")
        wiz = request.env['ca.registra_persona'].with_context(no_compute_tag_id_number=True).create(vals)
        wiz.action_confirm()
        if add_doc:
            return request.redirect(f"/badge_release_docs/{wiz.persona_id.id}")
        return request.redirect('/anagrafiche')

    @http.route('/get/badge_release/init_data', auth='user', type='json', website=True)
    def get_badge_release_init_data(self, **kwargs):
        user = request.env.user
        if not check_access_permission(user):
            raise Forbidden()
        
        Env = request.env
        
        tags = Env['ca.tag'].search_read(
            [('in_use', '=', False), ('revoked', '=', False)],
            fields=['id', 'name', 'display_name', 'tag_code', 'temp', 'ca_proprieta_tag_ids']
        )

        tag_domains = {
            'visitor': [Env.ref('inrim_anagrafiche.proprieta_tag_visitatore').id, Env.ref('inrim_anagrafiche.proprieta_tag_servizio').id],
            'internal': [Env.ref('inrim_anagrafiche.proprieta_tag_jolly').id, Env.ref('inrim_anagrafiche.proprieta_tag_definitivo').id],
            'temp': [Env.ref('inrim_anagrafiche.proprieta_tag_jolly').id]
        }
        
        work_infos = Env['ca.work_info'].search_read(
            [('state', '=', 'active')],
            fields=['ca_persona_id', 'date_start', 'date_end', 'ca_work_info_type_id', 'ca_title_id']
        )

        ids_hidden = [
            Env.ref('inrim_anagrafiche.tipo_ente_azienda_sede').id,
            Env.ref('inrim_anagrafiche.tipo_ente_azienda_sede_distaccata').id
        ]
        
        tipo_enti_hidden = Env['ca.tipo_ente_azienda'].search_read([('id', 'in', ids_hidden)], ['id', 'display_name'])
        
        tipo_enti = Env['ca.tipo_ente_azienda'].search_read(
            [('id', 'not in', ids_hidden)], 
            ['id', 'name', 'display_name']
        )

        return {
            'tags': tags,
            'tag_domains': tag_domains,
            'work_infos': work_infos,
            'titoli': Env['ca.titolo_persona'].search_read([], ['id', 'name', 'display_name', 'structured']),
            'work_info_types': Env['ca.work_info_type'].search_read([], ['id', 'name', 'display_name']),
            'enti_aziende': Env['ca.ente_azienda'].search_read([], ['id', 'name', 'display_name', 'vat', 'tipo_ente_azienda_id']),
            'tipo_enti_hidden_ids': [t['id'] for t in tipo_enti_hidden],
            'tipo_enti': tipo_enti,
            'persona_parent': Env['ca.persona'].search_read([('is_internal', '=', True)], ['id', 'display_name', 'present'])
        }

    @http.route('/get/badge_release/ca_persona', auth='user', type='json', website=True)
    def badge_release_ca_persona(self, **kwargs):
        user = request.env.user
        if not check_access_permission(user):
            raise Forbidden()
        return self._get_badge_release_personas(request.env)
